### PR TITLE
perf test ingestor

### PR DESCRIPTION
**Pull request by GeorgeJahad**
# Resolves

https://jira.rax.io/browse/SALUS-675

# What

Creates a perf test mode which allows the envoy to generate dummy metrics for performance testing.  Also includes a webserver that accepts rest calls to change the rate of metrics per minute and the size/(number of float values) in each metric sent.

# How

A new "--perf-test-port=<port number>" command line option invokes perf test mode and starts the webserver on <port number> parameter.

That webserver creates a rest endpoint which allows the rate/size of the metrics to be changed.

The metrics generation is implemented with a standalone "ingestor", i.e. one that doesn't ingest but just generates metrics on its own with connecting to an external agent.

## How to test

unit tests added

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -160,6 +160,16 @@ An example invocation of this sub-command is:
...
 
+### Running perf test mode
+```bash
+telemetry-envoy run --perf-test-port=8100 --config=envoy-config-provided.yml
+```
+starts the envoy in perf test mode at port 8100
+
+The rate of metrics generated can be changed with a rest call like so:
+```bash
+curl  "http://localhost:8100/?metricsPerMinute=15&floatsPerMetric=20"
```
Comment (itzg) [inline README.md]: It would be better to expose the endpoint as a `POST` where invoking curl like this will send the same query parameters in the body since the default is form url encoding:

```
curl http://localhost:8100 -d metricsPerMinute=15 -d floatsPerMetric=20
```

File: ingest/perf_test_ingestor.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+		case <-p.ticker.C:
+			p.processMetric()
+		}
+	}
+}
+
+func (p *PerfTestIngestor) startPerfTestServer() {
+	serverMux := http.NewServeMux()
+	serverMux.Handle("/", p.serverHandler)
```
Comment (itzg) [inline ingest/perf_test_ingestor.go]: If you use `serverMux.HandleFunc` I think you could pass `p.handler` directly rather than referencing via `p.serverHandler` field.

File: ingest/perf_test_ingestor.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+		return
+	}
+
+	go p.startPerfTestServer()
+	for {
+		if p.previousMetricsPerMinute != p.currentMetricsPerMinute {
+			if p.ticker != nil {
+				p.ticker.Stop()
+			}
+			p.previousMetricsPerMinute = p.currentMetricsPerMinute
```
Comment (itzg) [inline ingest/perf_test_ingestor.go]: This approach is totally fine, but an even Go-er way to do this could be to use a `chan int` and have the endpoint handler send the new rate to that channel and receive from the channel in the `select` below and do the ticker restart there.

File: ingest/perf_test_ingestor.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+func (p *PerfTestIngestor) handler(w http.ResponseWriter, r *http.Request) {
+	params := r.URL.Query()
+	metricsPerMinuteSet := false
+	floatsPerMetricSet := false
+
+	metricsPerMinuteVals, ok := params["metricsPerMinute"]
+	if ok && len(metricsPerMinuteVals) != 0 {
+		metricsCount, err := strconv.Atoi(metricsPerMinuteVals[0])
```
Comment (itzg) [inline ingest/perf_test_ingestor.go]: It would have been nice if https://golang.org/pkg/net/url/#Values.Get returned an `ok` indication along with the value. If you wanted to avoid the map-and-slice access you could do `if params.Get("metricsPerMinute") != "" {`

File: ingest/perf_test_ingestor.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+		if err != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			_, _ = w.Write([]byte("floatsPerMetric parameter must be an int: " + err.Error()))
+			return
+		}
+		p.floatsPerMetric = int64(floatsCount)
+		floatsPerMetricSet = true
+	}
+	if metricsPerMinuteSet == false && floatsPerMetricSet == false {
+		_, _ = w.Write([]byte("metricsPerMinute or floatsPerMetric parameter required"))
```
Comment (itzg) [inline ingest/perf_test_ingestor.go]: You might be missing a `w.WriteHeader(http.StatusBadRequest)` here unless a 200 OK was intended as a warning case.

File: ingest/perf_test_ingestor.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+func (p *PerfTestIngestor) processMetric() {
+	fvalues := make(map[string]float64)
+	svalues := make(map[string]string)
+	tags := make(map[string]string)
+	tags["test_tag"] = "perfTestTag"
+	svalues["result_type"] = "success"
+	for i := 0; int64(i) < p.floatsPerMetric; i++ {
+		fvalues[fmt.Sprintf("result_code%d", i)] = -2.0
```
Comment (itzg) [inline ingest/perf_test_ingestor.go]: Could vary the metric value yet keep it deterministic with a `float64(i)` as the value

File: README.md (hunk 1 above)
Comment (GeorgeJahad) [reply README.md]: I don't mind doing this at all, but just want to be sure I understand the reasoning.  Is it because this is sending info to the server and so a post is more semantically accurate than a get?

File: ingest/perf_test_ingestor.go (hunk 2 above)
Comment (GeorgeJahad) [reply ingest/perf_test_ingestor.go]: I had a problem with go recognizing it as a handler if I did it that way.  Let me try again.

File: ingest/perf_test_ingestor.go (hunk 3 above)
Comment (GeorgeJahad) [reply ingest/perf_test_ingestor.go]: that sounds much more elegant.  let me try that

File: ingest/perf_test_ingestor.go (hunk 5 above)
Comment (GeorgeJahad) [reply ingest/perf_test_ingestor.go]: good catch

File: README.md (hunk 1 above)
Comment (itzg) [reply README.md]: Sorry, I should have explained my logic. Yes, semantically a `GET` should be non-modifying.

File: ingest/perf_test_ingestor.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+		log.Fatalf("couldn't create perf test server")
+	}
+	log.Info("started perfTest webServer")
+	err = http.Serve(listener, serverMux)
+	log.Fatalf("perf test server error %v", err)
+}
+
+func (p *PerfTestIngestor) handler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		w.WriteHeader(http.StatusBadRequest)
```
Comment (itzg) [inline ingest/perf_test_ingestor.go]: There's actually a status code for that
```suggestion
		w.WriteHeader(http.StatusMethodNotAllowed)
```

File: ingest/perf_test_ingestor.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+		_, _ = w.Write([]byte("Only POST's accepted"))
+		return
+	}
+	params := r.URL.Query()
```
Comment (itzg) [inline ingest/perf_test_ingestor.go]: I was actually thinking the params would get parsed out of the POST body, which curl will send with url form encoding using:

https://golang.org/pkg/net/http/#Request.ParseForm

File: README.md [hunk 9]
```diff
@@ -160,6 +160,16 @@ An example invocation of this sub-command is:
...
 
+### Running perf test mode
+```bash
+telemetry-envoy run --perf-test-port=8100 --config=envoy-config-provided.yml
+```
+starts the envoy in perf test mode at port 8100
+
+The rate of metrics generated can be changed with a rest call like so:
+```bash
+curl -X POST  "http://localhost:8100/?metricsPerMinute=15&floatsPerMetric=20"
```
Comment (itzg) [inline README.md]: If you use the following curl will automatically assume a POST and set the content type to url form encoding:

```
curl http://localhost:8100/ -d metricsPerMinute=15 -d floatsPerMetric=20
```

File: ingest/perf_test_ingestor_test.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+	assert.Equal(t, float64(0),
+		args[0].Variant.(*telemetry_edge.Metric_NameTagValue).NameTagValue.Fvalues["result_code0"])
+	assert.Equal(t, 10,
+		len(args[0].Variant.(*telemetry_edge.Metric_NameTagValue).NameTagValue.Fvalues))
+	assert.Equal(t, "success",
+		args[0].Variant.(*telemetry_edge.Metric_NameTagValue).NameTagValue.Svalues["result_type"])
+
+	metricsPerMinute := 10
+	floatsPerMetric := 20
+	resp, err := http.Post(fmt.Sprintf("http://localhost:%d/?metricsPerMinute=%d&floatsPerMetric=%d", port, metricsPerMinute, floatsPerMetric), "", nil)
```
Comment (itzg) [inline ingest/perf_test_ingestor_test.go]: Sorry, my POST body suggestion is going to make this code require a little more setup

File: ingest/perf_test_ingestor.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+func (p *PerfTestIngestor) processMetric() {
+	fvalues := make(map[string]float64)
+	svalues := make(map[string]string)
+	tags := make(map[string]string)
+	tags["test_tag"] = "perfTestTag"
+	svalues["result_type"] = "success"
+	for i := 0; i < p.floatsPerMetric; i++ {
+		fvalues[fmt.Sprintf("result_code%d", i)] = float64(i)
```
Comment (itzg) [inline ingest/perf_test_ingestor.go]: total nit pick: a result code would probably have an integer value, how about
```suggestion
		fvalues[fmt.Sprintf("duration_%d", i)] = float64(i)
```

File: ingest/perf_test_ingestor.go [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+			p.ticker = time.NewTicker(time.Duration(int64(time.Minute) / int64(p.metricsPerMinute)))
+		case <-p.ticker.C:
+			p.processMetric()
+		}
+	}
+}
+
+func (p *PerfTestIngestor) startPerfTestServer() {
+	serverMux := http.NewServeMux()
+	serverMux.Handle("/", http.HandlerFunc(p.handler))
```
Comment (itzg) [inline ingest/perf_test_ingestor.go]: If you want to shave off a few characters, this seems to work
```suggestion
	serverMux.HandleFunc("/", p.handler)
```

however, I'm now confused why they also have `http.HandlerFunc`. I guess it's for middleware that accepts only `http.Handler` instances.

File: cloudbuild.yaml [hunk 13]
```diff
@@ -8,7 +8,7 @@ steps:
       - |
         set -ex
         make init
-        make test-report-junit
+        make test
         cat test-results/report.xml
```
Comment (itzg) [inline cloudbuild.yaml]: Mentioned in Slack too, but this line now needs to be removed.

**Comment by GeorgeJahad** [comment]
> Excellent. Thanks for accommodating my insistence on the POST changes :)

Happy to do it.  You are always teaching me new things.
